<a href="https://colab.research.google.com/github/MarinaVoit/python_for_ds_tasks/blob/main/Maryna_Voit_hw_10_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

import pandas as pd
df = pd.read_csv('drive/MyDrive/Colab/data/supermarket_sales.csv')

Mounted at /content/drive


1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [2]:
df.sort_values(by="Unit price", ascending=False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [10]:
el_acces = df[df['Product line'] == 'Electronic accessories']
el_acces.sort_values(by=['Quantity', 'Unit price'], ascending=False).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [19]:
df['Rating Category'] = df['Rating'].apply(
    lambda x: 'High' if x >= 8 else ('Medium' if x >= 6 else 'Low')
)

In [29]:
# def rating_category(rating):
#     if rating >= 8:
#         return 'High'
#     elif rating >= 6:
#         return 'Medium'
#     else:
#         return 'Low'
# df['Rating Category'] = df['Rating'].apply(rating_category)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [30]:
(df['Rating Category'].value_counts(normalize=True) * 100).round(2)

,proportion
Rating Category,
Medium,35.6
High,32.9
Low,31.5


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [43]:
time = '11:28'

In [32]:
# get_hour = lambda x: int(x.split(':')[0])
# get_hour(time)

11

In [44]:
def get_hour(time):
    return pd.Series(time.split(':')[0]).astype(int)[0]

result = get_hour(time)
print(result, type(result))

11 <class 'numpy.int64'>


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [45]:
df['Hour'] = df['Time'].apply(get_hour)
df[['Time', 'Hour']].head()

,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [51]:
df['Hour'].value_counts()

,count
Hour,
19,113
13,103
15,102
10,101
18,93
11,90
12,89
14,83
16,77


Найбільше покупок здійснюють о 19:00 (113 покупок), на другому місці — 13:00 (103 покупки), та на третьому — 15:00 (102 покупки).

Так, це цілком логічно та обґрунтовано поведінкою покупців:
19:00 : Це час закінчення стандартного робочого дня, коли люди повертаються додому та заходять у магазин за продуктами на вечерю або побутовими товарами.
13:00 (пік о 13-й годині): Збігається з традиційною обідньою перервою, коли покупці купують готову їжу, перекуски або роблять швидкі покупки протягом вільного часу.
15:00: Може відповідати часу післяобідньої перерви, поверненню дітей зі школи або кава-брейкам.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [52]:
df.groupby('City')['Total'].mean()

,Total
City,
Mandalay,319.872506
Naypyitaw,337.099715
Yangon,312.354031


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [53]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2)

,count,mean,min,max
City,,,,
Mandalay,332,319.87,18.64,1022.49
Naypyitaw,328,337.10,10.68,1042.65
Yangon,340,312.35,12.69,1039.29


7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [54]:
df.to_csv('supermarket_sales_enriched.csv', index=False)